sycl/doc/extensions/Assert/SYCL_INTEL_assert_exception.asciidoc

```diff
@@ -0,0 +1,109 @@
+= SYCL_INTEL_assert_exception
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+// This is necessary for asciidoc, but not for asciidoctor
+:cpp: C++
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.
+GitHub does not render image icons.
+
+This document describes an extension to rename device-specific kernel queries
+to better describe the operations performed.
+
+== Name Strings
+
++SYCL_INTEL_assert_exception+
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 3.
+
+== Overview
+
+== Modifications of SYCL 2020 Specification
+
+=== Change Section 4.13.2 Exception class interface
+
+Add enum member `assert` to the `errc` enum class:
+
+[source,c++,`sycl::kernel`,linenums]
+----
+assert = /* implementation defined */
+----
+
+==== Change table 136 Values of `errc` enum
+
+Add row `assert`:
+
+[width="40%",frame="topbot",options="header,footer"]
+|======================
+|Standard SYCL Error Codes |Description
+|`assert` | Assert failure had happened in device code during kernel execution
+|======================
+
+== Issues
+
+None.
+
+== Revision History
+
+[cols="5,15,15,70"]
+[grid="rows"]
+[options="header"]
+|========================================
+|Rev|Date|Author|Changes
+|1|2021-04-08|Sergey Kanaev|*Initial public working draft*
+|========================================
```

Comment:
I think this specification also need to describe the `assert` macro itself.  The core SYCL 2020 spec says that device code cannot call arbitrary functions from the C++ library, unless they are specifically allowed.  Therefore, this spec needs to say that it's OK to call `assert()` from device code.

I think the spec also need to describe the behavior of `assert()` because it's different from the host behavior.  I'd suggest something like this:

> The behavior of `assert()` in device code is similar to its behavior in host code.  If the asserted condition is false, a message is printed to `stderr` and then the program typically aborts.  The mechanism for aborting the program is different, though.  Whereas the host version calls `std::abort()`, the device version causes an asynchronous SYCL `exception` with the `errc::ext_oneapi_assert` error code to be thrown.  The application can catch this exception like any other asynchronous exception that is thrown from an executing kernel.
>
> It is unspecified whether a failing `assert()` returns to its caller before the kernel terminates.  If a failing call returns, the device code may need to continue execution without deadlocking for the assertion message to be printed or for the exception to be thrown.
>
> The contents of the exception's `e.what()` string is unspecified.  Since the assertion message is already printed to `stderr` by the time the exception is thrown, the `e.what()` string may not include the location of the failing assertion.
>
> As with the host version of `assert()`, an application can disable assertions in device code by defining the `NDEBUG` preprocessor macro at the point when `<sycl.hpp>` is included.

BTW, is that last sentence accurate?  Does `NDEBUG` need to be defined at the point when `<sycl.hpp>` is included or at the point when `<cassert>` is included?

QUESTION: Does it make sense for the DPC++ runtime to throw a `errc::ext_oneapi_assert` exception when it detects a failed assert?  Maybe we should call `std::abort()` instead?  That might be more consistent with host assertions.

Reply:
Assert macro extension is already documented here: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst and https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst

Reply:
Those documents don't describe any of the issues I list in my proposed text above.  These are all differences between the standard C++ version of assert and the version we plan to implement for device code, so I think they need to be described.

In addition, all the API information about calling assert from device code should be described in one extension document.  If we are creating a new extension document for assert, that document should also describe the `assert` macro itself.  If you want to remove duplication, I'd suggest removing the description of `assert` from "C-CXX-StandardLibrary" and add a pointer to this new extension.  Alternatively, do not add a new extension document for assert, and add all this information to "C-CXX-StandardLibrary".

BTW, I realize that the text I proposed does not explicitly say that device code can call assert.  Tweaking my proposed text:

> This extension adds the ability for device code to call the C++ `assert()` macro.  The behavior of `assert()` in device code is similar to its behavior in host code.  If the asserted condition is false, a message is printed to `stderr` and then the program typically aborts.  The mechanism for aborting the program is different, though.  Whereas the host version calls `std::abort()`, the device version causes an asynchronous SYCL `exception` with the `errc::ext_oneapi_assert` error code to be thrown.  The application can catch this exception like any other asynchronous exception that is thrown from an executing kernel.
>
> It is unspecified whether a failing `assert()` returns to its caller before the kernel terminates.  If a failing call returns, the device code may need to continue execution without deadlocking for the assertion message to be printed or for the exception to be thrown.
>
> The contents of the exception's `e.what()` string is unspecified.  Since the assertion message is already printed to `stderr` by the time the exception is thrown, the `e.what()` string may not include the location of the failing assertion.
>
> As with the host version of `assert()`, an application can disable assertions in device code by defining the `NDEBUG` preprocessor macro at the point when `<sycl.hpp>` is included.

Reply:
Added